### PR TITLE
Upgrade React and relevant packages to v17

### PR DESCRIPTION
**Pull request by nevyangelova**
#### Summary
This is a second PR aiming to solve the popup state issues with the previous react 17 update. Initially I thought we need to update react bootstrap as well to be able to merge this but i did find a workaround to solve the issues with our popovers.

The react-bootstrap update is even bigger than this and the PR would have been difficult to review. So let's put it as its own OKR.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-34187

#### Release Note

```release-note
Added an update to React and all related packages.
```

### CODE REVIEW

**Comment by devinbinnie** [comment]
Re-assigning my review to @hmhealey as he's probably better fit to assess issues with this upgrade. Happy to still look at it if you want me to though.

File: components/advanced_create_comment/advanced_create_comment.tsx
Comment (hmhealey) [inline components/advanced_create_comment/advanced_create_comment.tsx:421]: This should probably have been done as the functional version of setState before
```suggestion
        this.setState((state) => {showEmojiPicker: !state.showEmojiPicker});
```

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: .npm-upgrade.json
Comment (M-ZubairAhmed) [inline .npm-upgrade.json:61]: 0/5 on removing these fields from here altogether, since we batch updates in separate PRs anyways

File: components/advanced_create_comment/advanced_create_comment.tsx
Comment (M-ZubairAhmed) [inline components/advanced_create_comment/advanced_create_comment.tsx:419]: nit : Does this look ok?
```suggestion
e?.stopPropagation()
```

Comment (M-ZubairAhmed) [reply components/advanced_create_comment/advanced_create_comment.tsx:419]: as we dont do anything else with the `event` object anyways

Comment (M-ZubairAhmed) [reply components/advanced_create_comment/advanced_create_comment.tsx:419]: if you decide to change to the above, please do the changes in other places as well where we did 
```
if (e) {
            e.stopPropagation();
        }
```

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: .npm-upgrade.json
Comment (hmhealey) [reply .npm-upgrade.json:61]: Since React 18 is out and we aren't going to want to update these during the regular monthly updates, I think it makes sense to keep these here

**Comment by nevyangelova** [comment]
@hmhealey @M-ZubairAhmed pipeline is finally working, ready for re-review 🎉

File: .circleci/config.yml
Comment (hmhealey) [inline .circleci/config.yml:103]: I guess you had no luck on fixing this? I might see if I can figure out how to get CI to pass without this since this feels not great

File: Makefile
Comment (hmhealey) [inline Makefile:60]: Same here

File: packages/components/rollup.config.js
Comment (hmhealey) [inline packages/components/rollup.config.js:36]: @crspeller Do you happen to remember why you used rollup-plugin-ts over `@rollup/plugin-typescript` before?

File: package.json
Comment (hmhealey) [inline package.json:13]: What are these needed for now? It doesn't look like we're actually using them since I'd assume we'd have to update the babel config for that

File: package.json
Comment (hmhealey) [inline package.json:34]: Similarly, is this still needed?

File: package.json
Comment (hmhealey) [inline package.json:314]: Same here

File: .circleci/config.yml
Comment (hmhealey) [reply .circleci/config.yml:103]: Got it. Some of the dependencies aren't being cached between CircleCI steps properly. A fix for this is [here](https://github.com/mattermost/mattermost-webapp/commit/b4eca4cf3426a970c077dfdfc05cde7751f67c8a)

File: packages/components/rollup.config.js
Comment (M-ZubairAhmed) [inline packages/components/rollup.config.js:8]: This package now not used can be removed

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]
Only 1 small request from my side and rest as Harrison comments. I will unblock from my end

**Review by M-ZubairAhmed — APPROVED** [review]

File: packages/components/rollup.config.js
Comment (crspeller) [reply packages/components/rollup.config.js:36]: I believe the `@rollup/plugin-typescript` one doesn't work with babel. Is there some motivation for changing it? I think having babel in the pipeline will make migrating components easier.

**Review by crspeller — CHANGES_REQUESTED** [review]

**Comment by nevyangelova** [comment]
@crspeller the previous ts plugin wasn't compatible with our pipeline and was giving falsey linter issues. Adding this plugin fixes out pipeline issues and AFAIK it was the recommended one to use with rollup as the most supported one. I found an issue while researching but they seem to have closed it so I can't find and link to the discussion :/

**Review by crspeller — APPROVED** [review]
Checked it out. There does seem to be some issue with typescript checking in the dependencies. I am worried that it's not the plugin and that it's just swallowing valid errors. But not going to block the upgrade on this.

**Comment by nevyangelova** [comment]
Let me try reverting the change because I did update the pipeline cache so it might have resolved it!

File: packages/components/rollup.config.js
Comment (hmhealey) [reply packages/components/rollup.config.js:36]: Interesting. I would've thought babel would be its own explicit plugin rather than being implicitly included by another one.

I guess whether or not we need Babel depends on if we're ever including raw JS in this package or anything else that isn't covered by the TypeScript compiler. I haven't taken a look at Babel in a while to see what's actually still needed from it. I see that there's plugins for libraries (lodash and styled-components), some ES proposals that may have native support now, and something for hot loading. The lodash and styled-components ones at least seem like they'd be relevant in this package, so it's probably best to keep babel in the pipeline still.

@nevyangelova Is this change still needed now that we fixed the CircleCI issues? `@rollup/plugin-typescript` was involved there, so I wonder if we can revert this change

**Comment by nevyangelova** [comment]
@hmhealey it works for me locally. Ill spin another server

**Comment by nevyangelova** [comment]
@hmhealey nice catch! I managed to find some changes that need to be added when updating pdfjs. They aren't a lot but its better to keep it separately. I will revert those and open a follow up PR

**Comment by nevyangelova** [comment]
@hmhealey kind reminder for a review :)

**Review by hmhealey — APPROVED** [review]

**Comment by jgilliam17** [comment]
Thanks @nevyangelova 
few issues 
- User profile popovers tests are failing on cypress, reproduces when manually testing on the PR server.
- CRT tip button not found (e2e report)
- Color picker popover for custom theme in settings is failing to render
- Few tests with Draw plugin posting are failing
- Emoji picker is not opening while editing post 

See full report [here](https://mattermost-cypress-report.s3.amazonaws.com/233696-7587f2a-onprem-ent-webapp-pr-10836-mm-ee-test:7587f2a/mochawesome.html) 

Please update branch and I can re-run e2es. Thanks.

**Comment by jgilliam17** [comment]
Thanks @nevyangelova 
I'll share a test server with draw plugin in a DM (built with PR hash)
- Draw plugin is causing a crash
Screen Shot 2022-09-02 at 1 52 43 PM

- I think leaving the profile popover open until user clicks again on the profile image would be hard to discover for a lot of users, you could end up with something like this (not that you would want to 😃 )
Screen Shot 2022-09-02 at 1 48 25 PM
Screen Shot 2022-09-02 at 2 01 35 PM

- Just noting here that Thread footer Profile popover closes on outside click as expected (not an issue, observation that not all profile popovers behave the same) 
Screen Shot 2022-09-02 at 1 58 24 PM
- Member list popover does not remain open on click

https://user-images.githubusercontent.com/52937121/188211534-64a6e751-278a-4e9d-aa7a-89c1acb3b645.mov

**Comment by jgilliam17** [comment]
@nevyangelova can you please resolve the conflicts so I can trigger e2es. Thanks 🙂

**Comment by nevyangelova** [comment]
@jgilliam17 resolved, also I don't get any errors with draw plugin locally. @jespino updated the plugin yesterday as well so that might help the React integration on master. If there any any issues following we can perhaps investigate separately as they would need to be handled within the plugin code.

**Comment by jgilliam17** [comment]
Thanks for running e2es @nevyangelova 
I think we can disregard Session expiry failures on the [report,](https://mattermost-cypress-report.s3.amazonaws.com/239912-f0320a0-onprem-ent-webapp-pr-10836-mm-ee-test:f0320a0/mochawesome.html) but there are few tests that are failing to find elements, and also custom theme color picker tests that are not failing on master.
Current version of Draft plugin (v0.0.4) is not working on test server with react 17  cc @jespino for visibility

**Comment by hmhealey** [comment]
@jgilliam17 Since you'll be out leading up to Mattercon, would you be okay if we consider this approved if we can get those tests passing?

For the draw plugin, that might be acceptable to break since I don't know if anyone actually uses it in production as long as it doesn't outright crash the app (since I can't remember if that's the case)

**Comment by nevyangelova** [comment]
Also bear in mind we are updating the Draw plugin to 0.0.5 because it has the necessary dependency updates as well

**Comment by jgilliam17** [comment]
Thanks both.
@hmhealey Draw plug causes a white screen, not sure how many people use it on community, but it will most likely break after merging this PR. 
We can ask for this plugin to be disabled 0/5. 
Yes, let’s fix the test and move the PR along.

**Comment by hmhealey** [comment]
The plugin's actually been disabled on community for a while. I think it was turned off during a purge of unofficial plugins a while back

**Comment by nevyangelova** [comment]
@jgilliam17 also with this PR we will merge the dependency for the 0.0.5 version of the plugin so that it wont break on community even if the plugin is not disabled.

**Comment by nevyangelova** [comment]
@jgilliam17 I updated the only related failing tests. A lot of them seem flaky and i will re-run in this PR. I found two tests that are not related to this PR but are related to the Desktop app download prompt so i will work on them separately. Ticket filed - https://mattermost.atlassian.net/browse/MM-46986

**Review by jgilliam17 — APPROVED** [review]
Thanks Nevy for fixing the failed tests.
Adding approval based on our discussion and the fact that Draw plugin is currently disabled on community and will be updated to v0.0.5 soon.
I'll re-run the e2es for you, last cycle [failed](https://automation-dashboard.vercel.app/cycles/1cfa77bb-7a03-4020-b864-d7c62b5d3d2a) to start.

**Pull request merged**
